compiler-core/src/type_/expression.rs

```diff
@@ -741,6 +743,76 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
         })
     }
 
+    /// Checks for inefficient usage of `list.length` for checking for the empty list.
+    ///
+    /// If we find one of these usages, emit a warning to use `list.is_empty` instead.
+    fn check_for_inefficient_empty_list_check(
+        &mut self,
+        binop: BinOp,
+        left: &TypedExpr,
+        right: &TypedExpr,
+        location: SrcSpan,
+    ) {
+        // Check if we have a call expression on one side and an int literal on the other.
+        let (fun, value) = match (&left, &right) {
+            (TypedExpr::Call { fun, .. }, TypedExpr::Int { value, .. })
+            | (TypedExpr::Int { value, .. }, TypedExpr::Call { fun, .. }) => (fun, value),
+            _ => return,
+        };
+
+        // Check if the call expression is a module reference.
+        let (module_name, module_alias, label) = match fun.as_ref() {
+            TypedExpr::ModuleSelect {
+                module_name,
+                module_alias,
+                label,
+                ..
+            } => (module_name, module_alias, label),
+            _ => return,
+        };
+
+        // Check if we have a `list.length` call from `gleam/list`.
+        if module_name != "gleam/list" || label != "length" {
+            return;
+        }
+
+        let list_module = match self.environment.imported_modules.get(module_alias) {
+            Some((_, list_module)) => list_module,
+            None => return,
+        };
+
+        // TODO: Figure out how to simulate loading the `gleam_stdlib` package in the tests.
+        const STDLIB_MODULE_NAME: &str = if cfg!(test) {
+            // In the tests we currently declare all modules ourselves in a package
+            // named "thepackage".
+            //
+            // Therefore, we need to use that as the "stdlib" module name when running
+            // the tests if we want the warning to work as expected.s
+            "thepackage"
```

Comment:
This part here is kind of gross, since it means that our tests won't fully line up with reality.

One way we might be able to potentially get around this is reworking the test fixtures to be able to define modules that come from different packages. Not sure how difficult that will be.

@lpil Do you have any ideas on the best way to proceed here?

Reply:
Let's permit the package name to be specified in the test.

Reply:
We're now setting the package name to `gleam_stdlib` in the tests, to reflect reality.